src/Security/VoterV2/PersonalDataVoter.php

```diff
@@ -38,8 +41,30 @@ protected function voteOnAttribute(string $attribute, $subject, TokenInterface $
             return false;
         }
 
+        return match ($attribute) {
+            self::UPDATE => $this->canUpdate($user, $subject),
+            self::TOGGLE_VISIBILITY => $this->canToggleVisibility($user, $subject),
+            default => false,
+        };
+    }
+
+    private function canUpdate(User $user, DonneePersonnelle $subject): bool
+    {
         return $user->isBeneficiaire()
             ? $this->checker->isGranted('UPDATE', $subject->getBeneficiaire())
             : $this->checker->isGranted('UPDATE', $subject->getBeneficiaire()) && false === $subject->getBPrive();
     }
+
+    private function canToggleVisibility(User $user, DonneePersonnelle $subject): bool
+    {
+        $canUpdate = $this->canUpdate($user, $subject);
+
+        if ($subject instanceof Dossier) {
+            return !$subject->getDossierParent() && $canUpdate;
+        } elseif ($subject instanceof Document) {
+            return !$subject->getDossier() && $canUpdate;
+        }
+
+        return $canUpdate;
+    }
```

Comment:
I would create a method in DonneePersonnelle 
HasNoParent (or something similar with a better name) and override it in Dossier and Document. 
That way we could simply right this : 
```
private function canToggleVisibility(User $user, DonneePersonnelle $subject): bool
    {
        return $subject->hasNoParent() && $this->canUpdate($user, $subject);
    }
```

Reply:
I kind of forgot that DonneePersonnelle can also refers to Contact or else... 
Having a `hasNoParent()` method in it might not make sense. However, I do think that having an interface class for Dossier and Document would make a lot of sense.